### PR TITLE
Fix completion issue

### PR DESCRIPTION
**Pull request by casatir**
Completion implemented in #1155 is not fully working with JQuery terminal in `console.html`. It actually drops every completions that does not start with the prefix it computes (using whitespace splitting). See [this comment](https://github.com/iodide-project/pyodide/pull/1155#issuecomment-763981531).

One way to fix this is to prepend the prefix to each completion choice but this PR goes another way. CPython has a builtin completion system to work with `readline` called `rlcompleter` which you can use without standalone. This is the completion used in CPython's REPL. This will result in a faster completion experience, without any extra dependency (CPython's core instead of jedi/parso). Moreover, it already computes the completions with the prefix JQuery terminal is looking for.

This PR re-implements the completion in `InteractiveConsole` while fixing the issue exposed in [this comment](https://github.com/iodide-project/pyodide/pull/1155#issuecomment-763981531).

### CODE REVIEW

**Comment by rth** [comment]
So if I understand correctly, IPython is also using `rlcompleter` with an option use of jedi if it's installed? https://github.com/ipython/ipython/blob/master/IPython/core/completer.py#L68  Then `rlcompleter` should indeed be sufficient for us as well. jedi package is 1.4 MB compressed so it's clearly not negligible for web usage.

**Comment by casatir** [comment]
One little issue remains. Completion works only on prefixes preceded by a whitespace as JQuery Terminal split words at whitespaces.

```python
>>> str.isa <tab> <tab>
str.isalnum(        str.isalpha(        str.isascii(
>>> a = 5 ; str.isa <tab> <tab>
str.isalnum(        str.isalpha(        str.isascii(
>>> a = 5 ;str.isa <tab> <tab>  # no proposition
```

This is [known upstream](https://github.com/jcubic/jquery.terminal/issues/489) and could be fixed with `wordAutocomplete: false` and a bit of work. Should this be fixed in this PR? WDYT?

**Comment by rth** [comment]
> Should this be fixed in this PR? WDYT?

I think having this PR as is is already good. I'll try to review later today. We could always improve it in a follow up PR.

**Comment by casatir** [comment]
> So if I understand correctly, IPython is also using rlcompleter with an option use of jedi if it's installed? https://github.com/ipython/ipython/blob/master/IPython/core/completer.py#L68

Thanks, haven't noticed:

> This module started as fork of the rlcompleter module in the Python standard
library.  The original enhancements made to rlcompleter have been sent
upstream and were accepted as of Python 2.3,

> Then rlcompleter should indeed be sufficient for us as well. jedi package is 1.4 MB compressed so it's clearly not negligible for web usage.

Totally agree.

**Comment by casatir** [comment]
> I think having this PR as is is already good. I'll try to review later today. We could always improve it in a follow up PR.

Great!

**Comment by hoodmane** [comment]
By the way, a major concern with using jedi for autocompletions is that jedi can cause fatal errors #821. It isn't great if requesting a completion causes the Python runtime to segfault.

File: src/pyodide-py/pyodide/console.py [hunk 1]
```diff
@@ -250,38 +248,43 @@ def banner(self):
...
-
-        # here jedi should be ready
-        import jedi
-
-        return jedi.Interpreter(source, [self.locals]).complete()  # type: ignore
+        # not sure if this will be faster with regex like that:
+        # import re
+        # regex = '|'.join(map(re.escape, self.completer_word_break_characters))
+        # start = len(source) - re.search(regex, source[::-1]).start()
+        start = max(map(source.rfind, self.completer_word_break_characters)) + 1
```
[the remark this answers is not in the document]
Comment (casatir) [reply src/pyodide-py/pyodide/console.py:281]: Great! Thanks for the benchmark!

**Pull request merged**